shared-bindings/wiznet/__init__.c

```diff
@@ -0,0 +1,526 @@
+        *_errno = MP_EINVAL;
+        return MP_STREAM_ERROR;
+    }
+}
+
+#if 0
+STATIC void wiznet5k_socket_print(void (*print)(void *env, const char *fmt, ...), void *env, mp_obj_t self_in, mp_print_kind_t kind) {
+    wiznet5k_socket_obj_t *self = self_in;
+    print(env, "<WIZNET5K.socket sn=%u MR=0x%02x>", self->sn, getSn_MR(self->sn));
+}
+
+STATIC mp_obj_t wiznet5k_socket_disconnect(mp_obj_t self_in) {
+    mp_int_t ret = WIZCHIP_EXPORT(disconnect)(self->sn);
+    return 0;
+}
+#endif
+
+void create_random_mac_address(uint8_t *mac) {
+    uint32_t rb1 = shared_modules_random_getrandbits(24);
+    uint32_t rb2 = shared_modules_random_getrandbits(24);
+    // first octet has multicast bit (0) cleared and local bit (1) set
+    // everything else is just set randomly
+    mac[0] = ((uint8_t)(rb1 >> 16) & 0xfe) | 0x02;
+    mac[1] = (uint8_t)(rb1 >> 8);
+    mac[2] = (uint8_t)(rb1);
+    mac[3] = (uint8_t)(rb2 >> 16);
+    mac[4] = (uint8_t)(rb2 >> 8);
+    mac[5] = (uint8_t)(rb2);
+}
+
+/******************************************************************************/
+// MicroPython bindings
+
+/// \classmethod \constructor(spi, pin_cs, pin_rst)
+/// Create and return a WIZNET5K object.
+STATIC mp_obj_t wiznet5k_make_new(const mp_obj_type_t *type, size_t n_args, size_t n_kw, const mp_obj_t *args) {
+    // check arguments
+    mp_arg_check_num(n_args, n_kw, 3, 3, false);
+
+    // init the wiznet5k object
+    wiznet5k_obj.base.type = (mp_obj_type_t*)&mod_network_nic_type_wiznet5k;
+    wiznet5k_obj.cris_state = 0;
+    wiznet5k_obj.spi = MP_OBJ_TO_PTR(args[0]);
+    common_hal_digitalio_digitalinout_construct(&wiznet5k_obj.cs, args[1]);
+    common_hal_digitalio_digitalinout_construct(&wiznet5k_obj.rst, args[2]);
+    wiznet5k_obj.socket_used = 0;
+
+    /*!< SPI configuration */
+    // XXX probably should check if the provided SPI is already configured, and
+    // if so skip configuration?
+
+    common_hal_busio_spi_configure(wiznet5k_obj.spi,
+        10000000,  // BAUDRATE 10MHz
+        1, // HIGH POLARITY
+        1, // SECOND PHASE TRANSITION
+        8 // 8 BITS
+    );
+
+    common_hal_digitalio_digitalinout_switch_to_output(&wiznet5k_obj.cs, 1, DRIVE_MODE_PUSH_PULL);
+    common_hal_digitalio_digitalinout_switch_to_output(&wiznet5k_obj.rst, 1, DRIVE_MODE_PUSH_PULL); 
+
+    common_hal_digitalio_digitalinout_set_value(&wiznet5k_obj.rst, 0);
+    mp_hal_delay_us(10); // datasheet says 2us
+    common_hal_digitalio_digitalinout_set_value(&wiznet5k_obj.rst, 1);
+    mp_hal_delay_ms(160); // datasheet says 150ms
+
+    reg_wizchip_cris_cbfunc(wiz_cris_enter, wiz_cris_exit);
+    reg_wizchip_cs_cbfunc(wiz_cs_select, wiz_cs_deselect);
+    reg_wizchip_spi_cbfunc(wiz_spi_read, wiz_spi_write);
+
+    uint8_t sn_size[16] = {2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2}; // 2k buffer for each socket
+    ctlwizchip(CW_INIT_WIZCHIP, sn_size);
+
+    // set some sensible default values; they are configurable using ifconfig method
```

Comment:
Can we default to DHCP instead? My guess is that will be much more common. Maybe take ip, netmask, gateway and dns as kwargs instead in case static allocation is wanted instead.

Reply:
Yeah, if you'd rather.